openslides/utils/projector.py

```diff
@@ -1,66 +1,56 @@
-        """
-        self.projector_object = projector_object
-        self.config_entry = config_entry
-        assert self.config_entry.get('name') == self.name, (
-            'To get data of a projector element, the correct config entry has to be given.')
-        self.check_data()
-        return self.update_data() or {}
 
-    def check_data(self) -> None:
-        """
-        Method can be overridden to validate projector element data. This
-        may raise ProjectorException in case of an error.
+projector_elements: Dict[str, ProjectorElementCallable] = {}
 
-        Default: Does nothing.
-        """
-        pass
 
-    def update_data(self) -> Dict[Any, Any]:
-        """
-        Method can be overridden to update the projector element data
-        output. This should return a dictonary. Use this for server
-        calculated data which have to be forwared to the client.
+def register_projector_element(name: str, element: ProjectorElementCallable) -> None:
+    """
+    Registers a projector element.
 
-        Default: Does nothing.
-        """
-        pass
+    Has to be called in the app.ready method.
+    """
+    projector_elements[name] = element
 
 
-projector_elements: Dict[str, ProjectorElement] = {}
+async def get_projectot_data(projector_ids: List[int] = None) -> Dict[int, Dict[str, Any]]:
+    """
+    Callculates and returns the data for one or all projectors.
+    """
+    if projector_ids is None:
+        projector_ids = []
 
+    all_data = await element_cache.get_all_full_data_ordered()
+    projector_data: Dict[int, Dict[str, Dict[str, Any]]] = {}
 
-def register_projector_elements(elements: Generator[Type[ProjectorElement], None, None]) -> None:
-    """
-    Registers projector elements for later use.
+    for projector_id, projector in all_data.get('core/projector', {}).items():
+        if projector_ids and projector_id not in projector_ids:
+            # only render the projector in question.
+            continue
 
-    Has to be called in the app.ready method.
-    """
-    for AppProjectorElement in elements:
-        element = AppProjectorElement()
-        projector_elements[element.name] = element  # type: ignore
+        projector_data[projector_id] = {}
+        if not projector['config']:
+            projector_data[projector_id] = {'error': {'error': 'Projector has no config'}}
+            continue
+
+        for uuid, projector_config in projector['config'].items():
+            projector_element = projector_elements.get(projector_config['name'])
+            if projector_element is None:
+                projector_data[projector_id][uuid] = {'error': 'Projector element {} does not exist'.format(projector_config['name'])}
+            else:
+                projector_data[projector_id][uuid] = projector_element(projector_config, all_data)
+    return projector_data
 
 
-def get_all_projector_elements() -> Dict[str, ProjectorElement]:
+def get_config(all_data: AllData, key: str) -> Any:
```

Comment:
Please move this function to the tests or remove it.

Reply:
It will be needed in the 3.0 release. See the mail from emanuel, that internal agenda_items should be shown on the projector, when a specific config is set.

So I would like this function to stay in this pull request. We use it as soon as the slide functions are implemented.